File: sdkcode/javascript/UniTranslate.js
const crypto = require('crypto');

class UniTranslate {
    /**
     * Generate HMAC-SHA256 signature for the given parameters
     * @param {string} key Secret key
     * @param {Object} params Parameters to sign
     * @returns {string} Hex-encoded signature
     */
    static authEncrypt(key, params) {
        const data = this.sortMapToStr(params);
        return crypto
            .createHmac('sha256', key)
            .update(data)
            .digest('hex');
    }

    /**
     * Convert map to sorted string representation
     * @param {Object} data Input data
     * @returns {string} Sorted string representation
     */
    static sortMapToStr(data) {
        if (!data || Object.keys(data).length === 0) {
            return '';
        }

        const keys = Object.keys(data).sort();
        return keys.map((k, i) => {
            let value;
            if (Array.isArray(data[k])) {
                // Handle array
                value = '[' + data[k].map(item => {
                    if (item && typeof item === 'object' && !Array.isArray(item)) {
                        return '{' + this.sortMapToStr(item) + '}';
                    }
                    return String(item);
                }).join(',') + ']';
            } else if (data[k] && typeof data[k] === 'object' && !Array.isArray(data[k])) {
                // Handle nested map
                value = '{' + this.sortMapToStr(data[k]) + '}';
            } else {
                value = String(data[k] === null ? '' : data[k]);
            }
            return `${k}:${value}`;
        }).join('&');
    }
}

module.exports = UniTranslate;

File: sdkcode/csharp/UniTranslate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace UniTranslate
{
    public class UniTranslateSDK
    {
        /// <summary>
        /// Generate HMAC-SHA256 signature for the given parameters
        /// </summary>
        /// <param name="key">Secret key</param>
        /// <param name="params">Parameters to sign</param>
        /// <returns>Hex-encoded signature</returns>
        public static string AuthEncrypt(string key, Dictionary<string, object> @params)
        {
            string data = SortMapToStr(@params);
            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key)))
            {
                byte[] hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Convert map to sorted string representation
        /// </summary>
        /// <param name="data">Input data</param>
        /// <returns>Sorted string representation</returns>
        private static string SortMapToStr(Dictionary<string, object> data)
        {
            if (data == null || !data.Any())
            {
                return string.Empty;
            }

            var sortedKeys = data.Keys.OrderBy(k => k).ToList();
            var parts = new List<string>();

            foreach (var k in sortedKeys)
            {
                var v = data[k];
                string value;

                if (v is Dictionary<string, object> dict)
                {
                    // Handle nested map
                    value = "{" + SortMapToStr(dict) + "}";
                }
                else if (v is IList<object> list)
                {
                    // Handle array
                    var arrayParts = new List<string>();
                    foreach (var item in list)
                    {
                        if (item is Dictionary<string, object> itemDict)
                        {
                            arrayParts.Add("{" + SortMapToStr(itemDict) + "}");
                        }
                        else
                        {
                            arrayParts.Add(item?.ToString() ?? "");
                        }
                    }
                    value = "[" + string.Join(",", arrayParts) + "]";
                }
                else
                {
                    value = v?.ToString() ?? "";
                }

                parts.Add($"{k}:{value}");
            }

            return string.Join("&", parts);
        }
    }
}
